File: src/utils/audit-logger.ts
import { appendFile, mkdir, stat } from 'node:fs/promises';
import { existsSync } from 'node:fs';
import { dirname } from 'node:path';
import { getAuditLogPath } from '../storage/keyring.js';

type AuditEventType =
  | 'SECRET_USED'
  | 'SECRET_USED_EXEC'
  | 'SECRET_BLOCKED'
  | 'SECRET_REDACTED'
  | 'SECRET_ADDED'
  | 'SECRET_REMOVED'
  | 'SECRET_ROTATED';

interface BaseAuditEvent {
  type: AuditEventType;
  timestamp: string;
  secret: string;
}

interface SecretUsedEvent extends BaseAuditEvent {
  type: 'SECRET_USED';
  domain: string;
  method: string;
  status: number;
  durationMs: number;
}

interface SecretUsedExecEvent extends BaseAuditEvent {
  type: 'SECRET_USED_EXEC';
  command: string;
  exitCode: number;
  durationMs: number;
}

interface SecretBlockedEvent extends BaseAuditEvent {
  type: 'SECRET_BLOCKED';
  domain: string;
  reason: 'DOMAIN_NOT_ALLOWED' | 'PLACEMENT_NOT_ALLOWED' | 'COMMAND_NOT_ALLOWED';
}

interface SecretRedactedEvent extends BaseAuditEvent {
  type: 'SECRET_REDACTED';
  responseBytes: number;
  redactedCount: number;
}

interface SecretAddedEvent extends BaseAuditEvent {
  type: 'SECRET_ADDED';
  domains: string[];
  placements: string[];
}

interface SecretRemovedEvent extends BaseAuditEvent {
  type: 'SECRET_REMOVED';
}

interface SecretRotatedEvent extends BaseAuditEvent {
  type: 'SECRET_ROTATED';
  previousUses: number;
}

type AuditEvent =
  | SecretUsedEvent
  | SecretUsedExecEvent
  | SecretBlockedEvent
  | SecretRedactedEvent
  | SecretAddedEvent
  | SecretRemovedEvent
  | SecretRotatedEvent;

const MAX_LOG_SIZE = 10 * 1024 * 1024; // 10MB

async function ensureLogDir(): Promise<void> {
  const logPath = getAuditLogPath();
  const dir = dirname(logPath);
  if (!existsSync(dir)) {
    await mkdir(dir, { recursive: true, mode: 0o700 });
  }
}

async function shouldRotate(): Promise<boolean> {
  const logPath = getAuditLogPath();
  if (!existsSync(logPath)) return false;

  try {
    const stats = await stat(logPath);
    return stats.size >= MAX_LOG_SIZE;
  } catch {
    return false;
  }
}

async function rotateLog(): Promise<void> {
  const logPath = getAuditLogPath();
  const rotatedPath = `${logPath}.${Date.now()}.old`;

  const { rename } = await import('node:fs/promises');
  await rename(logPath, rotatedPath);
}

function formatEvent(event: AuditEvent): string {
  const parts = [`[${event.timestamp}]`, event.type, `secret=${event.secret}`];

  switch (event.type) {
    case 'SECRET_USED':
      parts.push(`domain=${event.domain}`, `method=${event.method}`, `status=${event.status}`, `duration=${event.durationMs}ms`);
      break;
    case 'SECRET_USED_EXEC':
      parts.push(`command=${event.command}`, `exit_code=${event.exitCode}`, `duration=${event.durationMs}ms`);
      break;
    case 'SECRET_BLOCKED':
      parts.push(`domain=${event.domain}`, `reason=${event.reason}`);
      break;
    case 'SECRET_REDACTED':
      parts.push(`response_bytes=${event.responseBytes}`, `redacted_count=${event.redactedCount}`);
      break;
    case 'SECRET_ADDED':
      parts.push(`domains=${event.domains.join(',')}`, `placements=${event.placements.join(',')}`);
      break;
    case 'SECRET_ROTATED':
      parts.push(`previous_uses=${event.previousUses}`);
      break;
  }

  return parts.join(' ');
}

async function log(event: AuditEvent): Promise<void> {
  await ensureLogDir();

  if (await shouldRotate()) {
    await rotateLog();
  }

  const logPath = getAuditLogPath();
  const line = formatEvent(event) + '\n';
  await appendFile(logPath, line, { mode: 0o600 });
}

export const audit = {
  async secretUsed(secret: string, domain: string, method: string, status: number, durationMs: number): Promise<void> {
    await log({
      type: 'SECRET_USED',
      timestamp: new Date().toISOString(),
      secret,
      domain,
      method,
      status,
      durationMs
    });
  },

  async secretUsedExec(secret: string, command: string, exitCode: number, durationMs: number): Promise<void> {
    await log({
      type: 'SECRET_USED_EXEC',
      timestamp: new Date().toISOString(),
      secret,
      command,
      exitCode,
      durationMs
    });
  },

  async secretBlocked(secret: string, domain: string, reason: 'DOMAIN_NOT_ALLOWED' | 'PLACEMENT_NOT_ALLOWED' | 'COMMAND_NOT_ALLOWED'): Promise<void> {
    await log({
      type: 'SECRET_BLOCKED',
      timestamp: new Date().toISOString(),
      secret,
      domain,
      reason
    });
  },

  async secretRedacted(secret: string, responseBytes: number, redactedCount: number): Promise<void> {
    await log({
      type: 'SECRET_REDACTED',
      timestamp: new Date().toISOString(),
      secret,
      responseBytes,
      redactedCount
    });
  },

  async secretAdded(secret: string, domains: string[], placements: string[]): Promise<void> {
    await log({
      type: 'SECRET_ADDED',
      timestamp: new Date().toISOString(),
      secret,
      domains,
      placements
    });
  },

  async secretRemoved(secret: string): Promise<void> {
    await log({
      type: 'SECRET_REMOVED',
      timestamp: new Date().toISOString(),
      secret
    });
  },

  async secretRotated(secret: string, previousUses: number): Promise<void> {
    await log({
      type: 'SECRET_ROTATED',
      timestamp: new Date().toISOString(),
      secret,
      previousUses
    });
  }
};

File: macos/Sources/CredentialProxyCore/Services/AuditLogger.swift
import Foundation

public enum AuditEventType: String, Codable {
    case SECRET_USED
    case SECRET_USED_EXEC
    case SECRET_BLOCKED
    case SECRET_REDACTED
    case SECRET_ADDED
    case SECRET_REMOVED
    case SECRET_ROTATED
}

public struct AuditEvent: Codable {
    public let type: AuditEventType
    public let timestamp: String
    public let secret: String

    // SECRET_USED / SECRET_BLOCKED
    public var domain: String?
    // SECRET_USED
    public var method: String?
    public var status: Int?
    // SECRET_USED / SECRET_USED_EXEC
    public var durationMs: Int?
    // SECRET_USED_EXEC
    public var command: String?
    public var exitCode: Int?
    // SECRET_BLOCKED
    public var reason: String?
    // SECRET_REDACTED
    public var responseBytes: Int?
    public var redactedCount: Int?
    // SECRET_ADDED
    public var domains: [String]?
    public var placements: [String]?
    // SECRET_ROTATED
    public var previousUses: Int?

    public init(
        type: AuditEventType,
        timestamp: String,
        secret: String,
        domain: String? = nil,
        method: String? = nil,
        status: Int? = nil,
        durationMs: Int? = nil,
        command: String? = nil,
        exitCode: Int? = nil,
        reason: String? = nil,
        responseBytes: Int? = nil,
        redactedCount: Int? = nil,
        domains: [String]? = nil,
        placements: [String]? = nil,
        previousUses: Int? = nil
    ) {
        self.type = type
        self.timestamp = timestamp
        self.secret = secret
        self.domain = domain
        self.method = method
        self.status = status
        self.durationMs = durationMs
        self.command = command
        self.exitCode = exitCode
        self.reason = reason
        self.responseBytes = responseBytes
        self.redactedCount = redactedCount
        self.domains = domains
        self.placements = placements
        self.previousUses = previousUses
    }
}

public class AuditLogger {
    private static let maxLogSize = 10 * 1024 * 1024 // 10MB

    public private(set) var logFilePath: String
    private let encoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.outputFormatting = .sortedKeys
        return encoder
    }()

    public init(logFilePath: String? = nil) {
        if let logFilePath {
            self.logFilePath = logFilePath
        } else {
            let appSupport = FileManager.default.urls(
                for: .applicationSupportDirectory, in: .userDomainMask
            ).first!.path
            self.logFilePath = "\(appSupport)/credential-proxy/audit.log"
        }
    }

    public func log(_ event: AuditEvent) {
        ensureLogDir()
        rotateIfNeeded()

        guard let data = try? encoder.encode(event),
              let json = String(data: data, encoding: .utf8) else {
            return
        }

        let line = json + "\n"
        let fileManager = FileManager.default

        if fileManager.fileExists(atPath: logFilePath) {
            guard let handle = FileHandle(forWritingAtPath: logFilePath) else { return }
            defer { handle.closeFile() }
            handle.seekToEndOfFile()
            handle.write(Data(line.utf8))
        } else {
            fileManager.createFile(
                atPath: logFilePath,
                contents: Data(line.utf8),
                attributes: [.posixPermissions: 0o600]
            )
        }
    }

    private func ensureLogDir() {
        let dir = (logFilePath as NSString).deletingLastPathComponent
        var isDir: ObjCBool = false
        if !FileManager.default.fileExists(atPath: dir, isDirectory: &isDir) {
            try? FileManager.default.createDirectory(
                atPath: dir,
                withIntermediateDirectories: true,
                attributes: [.posixPermissions: 0o700]
            )
        }
    }

    private func rotateIfNeeded() {
        let fileManager = FileManager.default
        guard fileManager.fileExists(atPath: logFilePath),
              let attrs = try? fileManager.attributesOfItem(atPath: logFilePath),
              let size = attrs[.size] as? Int,
              size >= Self.maxLogSize else {
            return
        }

        let timestamp = Int(Date().timeIntervalSince1970 * 1000)
        let rotatedPath = "\(logFilePath).\(timestamp).old"
        try? fileManager.moveItem(atPath: logFilePath, toPath: rotatedPath)
    }
}
